### PR TITLE
fix(metrics): separate the begin and view flow events

### PR DESCRIPTION
**Pull request by philbooth**
Related to mozilla/fxa-activity-metrics#19.

This PR makes two changes:

* `flow.${viewName}.begin` is renamed to `flow.begin` (yes, again).

* A new `flow.${viewName}.view` event is added.

The rationale is that we wanted to simplify the model for people writing queries in redash. And removing the view name from the `flow.begin` event makes it pleasingly symmetrical with the `flow.complete`, which is emitted by the auth server.

Adding the `flow.${viewName}.view` event enables us to more explicitly track which views any given flow travels through. And because it is emitted at the end of `afterVisible` its `flow_time` can be used as an approximation of the aggregated request time + initial client-side execution time, which may or may not be useful.

Here is an excerpt from the logs, showing what a flow looks like when a user lands on sign-in, then clicks on `Create account` to show the sign-up view, then clicks on `Have an account?` to go back to the sign-in view, then clicks on `Forgot password` to show the reset-password view:

```
{"event":"flow.begin","flow_id":"362c4127d4139f1303891b7dfe44cfe73eb7aa3c479bcae6ba38d866aa4ae94c","flow_time":0,"hostname":"pbooth-22205.local","op":"flowEvent","pid":34526,"time":"2016-11-23T12:27:45.739Z","userAgent":"Firefox-iOS-FxA/5.3 (Firefox)","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
{"event":"flow.signin.view","flow_id":"362c4127d4139f1303891b7dfe44cfe73eb7aa3c479bcae6ba38d866aa4ae94c","flow_time":721,"hostname":"pbooth-22205.local","op":"flowEvent","pid":34526,"time":"2016-11-23T12:27:46.460Z","userAgent":"Firefox-iOS-FxA/5.3 (Firefox)","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
{"event":"flow.signin.create-account","flow_id":"362c4127d4139f1303891b7dfe44cfe73eb7aa3c479bcae6ba38d866aa4ae94c","flow_time":10465,"hostname":"pbooth-22205.local","op":"flowEvent","pid":34526,"time":"2016-11-23T12:27:56.204Z","userAgent":"Firefox-iOS-FxA/5.3 (Firefox)","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
{"event":"flow.signup.view","flow_id":"362c4127d4139f1303891b7dfe44cfe73eb7aa3c479bcae6ba38d866aa4ae94c","flow_time":10521,"hostname":"pbooth-22205.local","op":"flowEvent","pid":34526,"time":"2016-11-23T12:27:56.260Z","userAgent":"Firefox-iOS-FxA/5.3 (Firefox)","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
{"event":"flow.signup.have-account","flow_id":"362c4127d4139f1303891b7dfe44cfe73eb7aa3c479bcae6ba38d866aa4ae94c","flow_time":19411,"hostname":"pbooth-22205.local","op":"flowEvent","pid":34526,"time":"2016-11-23T12:28:05.150Z","userAgent":"Firefox-iOS-FxA/5.3 (Firefox)","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
{"event":"flow.signin.view","flow_id":"362c4127d4139f1303891b7dfe44cfe73eb7aa3c479bcae6ba38d866aa4ae94c","flow_time":19450,"hostname":"pbooth-22205.local","op":"flowEvent","pid":34526,"time":"2016-11-23T12:28:05.189Z","userAgent":"Firefox-iOS-FxA/5.3 (Firefox)","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
{"event":"flow.signin.forgot-password","flow_id":"362c4127d4139f1303891b7dfe44cfe73eb7aa3c479bcae6ba38d866aa4ae94c","flow_time":23778,"hostname":"pbooth-22205.local","op":"flowEvent","pid":34526,"time":"2016-11-23T12:28:09.517Z","userAgent":"Firefox-iOS-FxA/5.3 (Firefox)","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
{"event":"flow.reset-password.view","flow_id":"362c4127d4139f1303891b7dfe44cfe73eb7aa3c479bcae6ba38d866aa4ae94c","flow_time":23796,"hostname":"pbooth-22205.local","op":"flowEvent","pid":34526,"time":"2016-11-23T12:28:09.535Z","userAgent":"Firefox-iOS-FxA/5.3 (Firefox)","v":1,"context":"fx_desktop_v3","entrypoint":"menupanel","service":"sync"}
```

@shane-tomlinson r?

### CODE REVIEW

**Comment by philbooth** [comment]
Thanks for the feedback @shane-tomlinson. I've tried to implement two of your suggestions in 07218ab:

* `flow.${viewName}.view` is now generated server-side, driven by the screen events.

* I moved the call to `logView` in `views/app.js` to immediately after the view is written to the DOM.

Can I get another r?

File: app/scripts/views/app.js [hunk 1]
```diff
@@ -122,6 +116,13 @@ define(function (require, exports, module) {
 
             this.writeToDOM(viewToShow.el);
 
+            // logView is done outside of the view itself because the settings
+            // page renders many child views at once. If the view took care of
+            // logging itself, each child view would be logged at the same time.
+            // We only want to log the screen being displayed, child views will
+            // be logged when they are opened.
+            viewToShow.logView();
```
Comment (shane-tomlinson) [inline app/scripts/views/app.js]: 👍 - I think this moving `logView` here is a good change.

**Comment by rfk** [comment]
> flow.${viewName}.begin is renamed to flow.begin (yes, again).

FWIW, one other potential benefit of this, is that we could one day log the `flow.begin` event on the server when they first request the page.  This would help us capture all the flows that abort before even waiting for all the resources to download.

**Comment by philbooth** [comment]
> ...we could one day log the `flow.begin` event on the server when they first request the page.

Although that would mean we also emit `flow.begin` for users that are already signed in. Which may be fine, I don't know, but it's different to what happens at the moment.

**Pull request merged**